CMakeLists.txt

```diff
@@ -416,86 +263,118 @@ if (S2N_INTERN_LIBCRYPTO)
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_LIBCRYPTO_SUPPORTS_EVP_MD_CTX_SET_PKEY_CTX)
-endif()
+    # Set the result of the probe
+    feature_probe_result(${PROBE_NAME} ${IS_AVAILABLE})
 
-# Determine if KEM Kyber512 implementation from AWS-LC is available
-try_compile(
-        LIBCRYPTO_SUPPORTS_EVP_KEM_KYBER_512
-        ${CMAKE_BINARY_DIR}
-        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/evp_kem_kyber_512.c"
-        LINK_LIBRARIES ${LINK_LIB} ${OS_LIBS}
-        COMPILE_DEFINITIONS "-Werror"
-)
+    # Make sure the variable is set in the parent scope
+    set(${PROBE_NAME} ${IS_AVAILABLE} PARENT_SCOPE)
 
-if(LIBCRYPTO_SUPPORTS_EVP_KEM_KYBER_512)
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_LIBCRYPTO_SUPPORTS_KYBER512)
-endif()
+    # Set the flags that we used for the probe
+    set(${PROBE_NAME}_FLAGS ${PROBE_FLAGS} PARENT_SCOPE)
+endfunction()
 
-# Determine if kTLS is supported
-try_compile(
-        PLATFORM_SUPPORTS_KTLS
-        ${CMAKE_BINARY_DIR}
-        SOURCES "${CMAKE_CURRENT_LIST_DIR}/tests/features/ktls.c"
-)
-if(PLATFORM_SUPPORTS_KTLS)
-    message(STATUS "feature probe: kTLS supported")
-    target_compile_options(${PROJECT_NAME} PUBLIC -DS2N_PLATFORM_SUPPORTS_KTLS)
-endif()
+# Iterate over all of the features and try to compile them
+FILE(GLOB FEATURE_SRCS "${CMAKE_CURRENT_LIST_DIR}/tests/features/*.c")
+list(SORT FEATURE_SRCS)
+foreach(file ${FEATURE_SRCS})
+    get_filename_component(feature_name ${file} NAME_WE)
+    feature_probe(${feature_name})
+endforeach()
 
-if (NOT DEFINED CMAKE_AR)
-	message(STATUS "CMAKE_AR undefined, setting to `ar` by default")
-	SET(CMAKE_AR ar)
-else()
-	message(STATUS "CMAKE_AR found: ${CMAKE_AR}")
+# FreeBSD might need to link to execinfo explicitly
+if(NOT S2N_EXECINFO_AVAILABLE AND CMAKE_SYSTEM_NAME STREQUAL "FreeBSD")
+    feature_probe(S2N_EXECINFO_AVAILABLE LINK_LIBRARIES execinfo)
 endif()
 
-if (NOT DEFINED CMAKE_RANLIB)
-    message(STATUS "CMAKE_RANLIB undefined, setting to `ranlib` by default")
-    SET(CMAKE_RANLIB ranlib)
-else()
-    message(STATUS "CMAKE_RANLIB found: ${CMAKE_RANLIB}")
+# Stack traces are only available if execinfo is
+if (NOT S2N_EXECINFO_AVAILABLE)
+    set(S2N_STACKTRACE FALSE)
 endif()
+feature_probe_result(S2N_STACKTRACE ${S2N_STACKTRACE})
 
-if (NOT DEFINED CMAKE_OBJCOPY)
-    message(STATUS "CMAKE_OBJCOPY undefined, setting to `objcopy` by default")
-    SET(CMAKE_OBJCOPY objcopy)
-else()
-    message(STATUS "CMAKE_OBJCOPY found: ${CMAKE_OBJCOPY}")
+set(S2N_KYBER512R3_AVX2_BMI2 FALSE)
+if(NOT S2N_NO_PQ_ASM)
+    # Kyber Round-3 code has several different optimizations which require
+    # specific compiler flags to be supported by the compiler.
+    # So for each needed instruction set extension we check if the compiler
+    # supports it and set proper compiler flags to be added later to the
+    # Kyber compilation units.
+    if(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^(x86_64|amd64|AMD64)$")
```

Comment:
Wouldn't you expect the compilation attempt to fail if the flags aren't supported? Or are they just ignored?